bin/desi_proc_joint_fit

```diff
@@ -366,13 +367,30 @@ if args.obstype in ['SCIENCE']:
         comm.barrier()
 
     if rank==0 and len(args.expids) > 1:
+        start_directory = os.path.abspath(os.curdir)
         for sp in spectro_nums:
             saved_stdfile = findfile('stdstars', args.night, args.expids[0], spectrograph=sp)
+            existing_dirname, existing_fname = os.path.split(saved_stdfile)
             for expid in args.expids[1:]:
                 new_stdfile = findfile('stdstars', args.night, expid, spectrograph=sp)
-                cmd = f'ln -s {saved_stdfile} {new_stdfile}'
-                log.info(f'Sym Linking jointly fitted stdstar file: {saved_stdfile}  to  {new_stdfile}')
+                new_dirname, new_fname = os.path.split(new_stdfile)
+                log.debug("Path exists: {}, file exists: {}, link exists: {}".format(os.path.exists(new_stdfile),
+                                                                                     os.path.isfile(new_stdfile),
+                                                                                     os.path.islink(new_stdfile)))
+                os.chdir(new_dirname)
+                rel_path_to_orig = os.path.relpath(existing_dirname,new_dirname)
+                rel_path_to_orig_name = os.path.join(rel_path_to_orig,existing_fname)
+                log.info(f'Sym Linking jointly fitted stdstar file: {new_stdfile} '+\
+                         f'to existing file at rel. path {rel_path_to_orig_name}')
+                cmd = f'ln -s {rel_path_to_orig_name} {new_fname}'
+                cmdlist = cmd.split(' ')
+                log.info(str(cmdlist))
```

Comment:
I think this logic is right, but you might be doing more directory parsing and command spawning than needed.  i.e. would this be equivalent but simpler for future maintenance?
```
relpath_saved_std = os.path.relpath(saved_stdfile, new_stdfile)
os.symlink(relpath_saved_std, new_stdfile)
```
(double check my src/dst order, but I think I go that right)

Related: what should code do if `saved_stdfile` doesn't exist (RuntimeError?), or if `new_stdfile` link already exists (replace it? log and move on?)

Reply:
I'm okay with making your change. I personally felt the opposite way about less code vs maintainable in this particular case, which is why I went with the longer solution. Your method works but I felt that the choices of keeping the first filename extension and throwing the second was rather arbitrary and I worried about the persistence of that choice in an external dependency. Maybe that was overly cautious. To illustrate my point:

Example:
```
os.path.relpath("./test1/foo.fits","./test2/bar.fits")
Out: '../../test1/foo.fits'
```

So it gives what we want, which is a path from the second directory to the first file. But I don't find it obvious that this should return the filename for the first entry and ignore the filename of the second entry. Because of that, I didn't think it was readably as clear or free from future changes in the os module.

Again, happy to change it though.


For the related point:
I purposefully give it to desispec's `runcmd` so that it is handled like any other function call within desi_proc. If the outputs exist, I believe that it skips them with a message stating that. If the input doesn't exist, it raises an error stating that. I personally think that is appropriate, but I'm happy to hear arguments to the contrary.

Reply:
Actually, your snippet doesn't exactly work. Now that I'm testing, I realize why I didn't choose to move forward with this. If you look at my toy example, it gives one too many ".." because it thinks that the file in the second argument is another directory to be traversed. So to make it work, you'd need to split off the directory for the second argument.

Similarly, I think we still need to change into the proper directory and use just the filename for the second argument in sym link, otherwise I don't think the relative path's will evaluate properly. I'm still testing this statement, but that's my naive thinking.

So to;dr this could be simplified slightly, but only a line or two and I think those save lines would hurt readability. 

The discussion on using runcmd vs os.symlink is still an open one, however. I'll wait for you to make an executive decision on that

Reply:
Thanks for the explanations.  Let's split the difference with
```
rel_path_to_orig_name = os.path.relpath(saved_stdfile, new_stdfile)
cmd = f'ln -s {rel_path_to_orig_name} {new_stdfile}'
runcmd(cmd, inputs=[saved_stdfile], outputs=[new_stdfile])
```
which minimizes path manipulation logic while providing the runcmd checks for what to do if the output already exists or if the input is missing.

Reply:
Hold the presses here.  Why is it necessary to call out to an external command to create the symlink?  I did not follow the logic of the discussion above.

Reply:
It's not necessary, runcmd just provides useful I/O handling which I argued makes it more convenient than os.symlink. I appreciate the fact that using os.symlink is more pythonic and would more smoothly work with the file system, but do you have reason to think this is a terrible idea?

If it will crash and burn, I can write my own I/O checks and use the os module.

Reply:
Yes, it is always a terrible idea to call out to the shell to do something you can do within the language itself.  I am not at all convinced that you can't do the IO handling in Python itself.  Could you describe in detail what you mean by "IO handling"?

Reply:
Also, how does this work with your unit tests?  I think it would be easier to construct a unit test around `os.symlink` instead of a system call.

Reply:
The motivation for using `runcmd` is that it provides some standardized checks and logging for what to do if a requested input is missing, what to do if the output already exists, and what to do if the command is run and returns but the output still doesn't exist.

However, @weaverba137 question reminded me that `runcmd` also accepts a callable with args as input instead of a string command to spawn.  I think that's the right solution here for this simple command.  Something like:
```
runcmd(os.symlink, args=(rel_path_to_orig_name, new_fname), \
    inputs=[rel_path_to_orig_name,], outputs=[new_fname,])
```

At the same time, we're spinning off to bikeshed territory in the name of maintainability (apologies for starting that...), so I'll leave it to @akremin to wrap this up with something that actually produces relative symlinks by whatever method, and we can revisit that implementation only if it becomes a problem in the future.

For the record: in other cases we preferred to make an actual command string that could be run from the command line so there is a super-direct connection between what is printed in the logging and what to run for debugging if it fails and so that any issues like memory leaks or leftover file handles are isolated in a process that runs and ends, but on some systems MPI+fork() may not be so happy about that so we may have to revisit it (in a different PR!)

Reply:
I will implement those checks in python and switch to using os.symlink.

For I/O, I am just referring to checking if the input files exist and output file exists, and raising the appropriate warnings/errors if they do/don't. It isn't hard to implement in python, I was just avoiding writing some code that existed in a useful plug-and-play way within the desispec ecosystem. It also had the benefit of being the same way we check all other files on disk when running our various functions, and how we execute the other function calls.

But you're right, using the native tool is certainly smarter.

Reply:
I've have used Stephen's compromise of using runcmd for consistent handling of the data while using the os version of symbolic linking

Reply:
To be clear, it now uses: 
```
runcmd(os.symlink, args=(relpath_saved_std, new_stdfile), \
       inputs=[relpath_saved_std, ], outputs=[new_stdfile, ])
```
Which means we are using the built-in os.symlink